### PR TITLE
Add support for JSON Config UI

### PR DESCRIPTION
**Pull request by UncleSamSwiss**
**PR Checklist:**

-   [x] Provide a meaningful description to this PR or mention which issues this fixes.
-   [x] Add tests for your change. This includes negative tests (i.e. inputs that need to fail) as well as baseline tests (i.e. how should the directory structure look like?).
-   [x] Run the test suite with `npm test`
-   [x] If there are baseline changes, review them and make a separate commit for them with the comment "accept baselines" if they are desired changes
-   [x] Ensure the project builds with `npm run build`
-   [x] If you added a required option, also add it to the template creation (`.github/create_templates.ts`)
-   [ ] Add a detailed migration description to `docs/updates` explaining what the user needs to do when manually updating an existing project
-   [x] Add your changes to `CHANGELOG.md` (referencing the migration description and this PR or the issue you fixed)

**Description:**  
For the admin UI the developer can now choose between:
- JSON UI
- HTML
- React
- No UI

### CODE REVIEW

File: src/lib/core/questions.ts [hunk 1]
```diff
@@ -552,24 +553,55 @@ export const questionGroups: QuestionGroup[] = [
...
+								".jsx",
+								(f) => !f.endsWith("tab.jsx"),
+						  )) ||
+						  (await ctx.hasFilesWithExtension(
+								"admin/src",
+								".tsx",
+								(f) => !f.endsWith("tab.tsx"),
+						  ))
+						? "react"
+						: "html",
```
Comment (AlCalzone) [inline src/lib/core/questions.ts]: I'd probably turn this into a proper function with if/else and return statements.

File: templates/admin/jsonConfig.json.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	if (!isAdapter || !useJsonConfig) return;
+
+	const adapterSettings: AdapterSettings[] = answers.adapterSettings ?? getDefaultAnswer("adapterSettings")!;
+	const items = adapterSettings.reduce<Record<string, any>>(
+		(old, setting) => ({... old, [setting.key]: generateSetting(setting)}), {})
+	const config = {
+		i18n: true,
+		type: "panel",
+		items
+	};
```
Comment (AlCalzone) [inline templates/admin/jsonConfig.json.ts:34]: I'm gonna trust you on this

File: templates/admin/style.css.ts [hunk 3]
```diff
@@ -3,7 +3,9 @@ import { readFile, TemplateFunction } from "../../src/lib/createAdapter";
 export = (answers => {
 
 	const isAdapter = answers.features.indexOf("adapter") > -1;
-	if (!isAdapter) return;
+	const useJsonConfig = answers.adminUi === 'json';
+	const noConfig = answers.adminUi === 'none';
+	if (!isAdapter || useJsonConfig || noConfig) return;
```
Comment (AlCalzone) [inline templates/admin/style.css.ts:8]: I think we can simplify this to html only, right?

File: templates/io-package.json.ts [hunk 4]
```diff
@@ -106,9 +124,12 @@ export = (async answers => {
 		`) : ""}
 		${supportCustom ? `"supportCustoms": true,` : ""}
 		"dependencies": [
-			${isAdapter ? `{ "js-controller": ">=2.0.0" },` : ""}
+			${isAdapter ? `{ "js-controller": ">=3.3.22" },` : ""}
 			${isWidget ? `"vis",` : ""}
 		],
+		"globalDependencies": [
+			${isAdapter ? `{ "admin": "${useJsonConfig ? ">=5.1.13" : ">=5.0.0" }" },` : ""}
+        ],
```
Comment (AlCalzone) [inline templates/io-package.json.ts]: turned some tabs into spaces here

File: test/baselines/TS_Prettier/package.json [hunk 5]
```diff
@@ -19,33 +19,33 @@
...
     "chai": "^4.3.6",
     "chai-as-promised": "^7.1.1",
-    "eslint": "^8.9.0",
-    "eslint-config-prettier": "^8.3.0",
+    "eslint": "^8.15.0",
+    "eslint-config-prettier": "^8.5.0",
     "eslint-plugin-prettier": "^4.0.0",
-    "mocha": "^9.2.0",
-    "prettier": "^2.5.1",
+    "mocha": "^10.0.0",
```
Comment (AlCalzone) [inline test/baselines/TS_Prettier/package.json]: mocha 10 has no node 12 support. should we pin it?

File: test/baselines/TS_Prettier/package.json [hunk 6]
```diff
@@ -19,33 +19,33 @@
...
+    "prettier": "^2.6.2",
     "proxyquire": "^2.1.3",
     "rimraf": "^3.0.2",
-    "sinon": "^13.0.1",
+    "sinon": "^14.0.0",
```
Comment (AlCalzone) [inline test/baselines/TS_Prettier/package.json]: sinon 14 has no node 12 support. should we pin it?

File: test/create-adapter.test.ts [hunk 7]
```diff
@@ -249,15 +249,16 @@ describe("adapter creation =>", () => {
...
 					language: "JavaScript",
 					es6class: "yes",
-					adminReact: "yes",
+					adminUi: "react",
 				};
 				await expectSuccess("adapter_JS_React", answers);
 			});
 
-			it("Adapter, JavaScript, ESLint, Spaces, Single quotes, Apache License", async () => {
+			it("Adapter, JavaScript, JSON UI, ESLint, Spaces, Single quotes, Apache License", async () => {
```
Comment (AlCalzone) [inline test/create-adapter.test.ts:257]: I'd mention the JSON UI in the test name so it appears in the directory name, line 268

**Review by AlCalzone — CHANGES_REQUESTED** [review]
Looks good, just a few details

File: templates/admin/style.css.ts (hunk 3 above)
Comment (UncleSamSwiss) [reply templates/admin/style.css.ts:8]: HTML and React

File: test/baselines/TS_Prettier/package.json (hunk 5 above)
Comment (UncleSamSwiss) [reply test/baselines/TS_Prettier/package.json]: Can you do this in #910 where the user can choose the node version?

File: test/baselines/TS_Prettier/package.json (hunk 6 above)
Comment (UncleSamSwiss) [reply test/baselines/TS_Prettier/package.json]: Can you do this in #910 where the user can choose the node version?

File: test/baselines/TS_Prettier/package.json (hunk 5 above)
Comment (AlCalzone) [reply test/baselines/TS_Prettier/package.json]: Actually I think I already do.

File: templates/admin/jsonConfig.json.ts (hunk 2 above)
Comment (UncleSamSwiss) [reply templates/admin/jsonConfig.json.ts:34]: Just tested it, looks good:

image

**Comment by UncleSamSwiss** [comment]
@AlCalzone can you explain me why I get those `Cannot find name 'systemDictionary'.ts(2304)` with my changes, but the code works fine with `master`?

**Comment by UncleSamSwiss** [comment]
> @AlCalzone can you explain me why I get those `Cannot find name 'systemDictionary'.ts(2304)` with my changes, but the code works fine with `master`?

OK, hopefully fixed it with the latest commit

File: templates/admin/jsonConfig.json.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { AdapterSettings, getDefaultAnswer } from "../../src/lib/core/questions";
+import type { TemplateFunction } from "../../src/lib/createAdapter";
+
+function generateSetting(settings: AdapterSettings): any {
+	if (settings.inputType === "select") {
+		return {
+			type: "select",
+			label: settings.label || settings.key,
+			options: settings.options.map(o => ({ label: o.text, value: o.value })),
+            newLine: true
```
Comment (AlCalzone) [inline templates/admin/jsonConfig.json.ts]:
```suggestion
			options: settings.options.map(o => ({ label: o.text, value: o.value })),
			newLine: true
```

File: templates/admin/jsonConfig.json.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+		};
+	} else {
+		return {
+			type: settings.inputType,
+			label: settings.label || settings.key,
+            newLine: true
```
Comment (AlCalzone) [inline templates/admin/jsonConfig.json.ts]:
```suggestion
			label: settings.label || settings.key,
			newLine: true
```

**Review by AlCalzone — CHANGES_REQUESTED** [review]

**Review by AlCalzone — APPROVED** [review]

**Pull request merged**